### ISSUE TITLE
Automatically install setuptools_dso

### ISSUE THREAD

**Issue by mtikekar** [mdavidsaver/setuptools_dso#12]
Currently, to install a package that uses setuptools_dso, we need to do:

```
pip install setuptools_dso
pip install package
```

But setuptools allows you to specify `setup_requires` which can install setuptools_dso automatically so that you only need to run `pip install package` (or `python setup.py install`). The catch is that setup.py should not import setuptools_dso. I found a nice example of how to do this in a package called CFFI (under the pypy project). This is adapted from their [docs](https://cffi.readthedocs.io/en/latest/overview.html):

```python
# in a file named piapprox_build.py in the same folder as setup.py
from cffi import FFI
ffibuilder = FFI()
ffibuilder.cdef(...)
ffibuilder.set_source(...)
```

```python
# in setup.py
from setuptools import setup

setup(
    ...
    setup_requires=["cffi>=1.0.0"],
    cffi_modules=["piapprox_build:ffibuilder"], # "filename:global"
    install_requires=["cffi>=1.0.0"],
)
```

By splitting the CFFI code into a separate file and referencing the CFFI objects as strings in setup.py, it allows setuptools to install CFFI before running the code that requires it.

Would it be interesting/possible to add this to setuptools_dso? You can continue to support the current way of doing things i.e. provide `ext_modules` as a list of Extension objects and also provide `setuptools_dso_modules` as a list of strings.

This is the CFFI [source repo](https://foss.heptapod.net/pypy/cffi). I haven't figured out how they plug into setuptools.

**Comment by navytux** [comment]
I would say today the build-time requirements are better specified via pyproject.toml. The reason for this is that else there is a cyclic dependency: say setup.py indicates via setup_requires that it needs setuptools_dso. And at the same time setup.py needs to import setuptools_dso -> it is not possible to execute setup.py at all. The pyproject.toml approach allows to avoid that. Here is how e.g. pygolang build-requires setuptools_dso:

```
[build-system]
requires = ["setuptools", "wheel", "setuptools_dso >= 1.7", "cython", "gevent"]
```
(https://lab.nexedi.com/nexedi/pygolang/blob/73fdea50/pyproject.toml)

Please see [PEP 518](https://www.python.org/dev/peps/pep-0518/) for more details.

In other words there is nothing on setuptools_dso to do: if a project puts setuptools_dso into pyproject.toml, pip install that project will automatically preinstall setuptools_dso at build time.

**Comment by mtikekar** [comment]
Yes, pyproject.toml would be one way of doing it and requires nothing from setuptools_dso to support. I was simply pointing out another way of avoiding the cyclic dependency that has been available since before pyproject.toml came along.

**Comment by mdavidsaver** [comment]
I would also recommend use of pyproject.toml to express "build time" dependencies.  If memory serves me, it was introduced with PIP version 10.

That said, I don't see any barrier to your (@mtikekar) adding something analogous to `cffi_modules` (I guess `dso_modules`?).  As you say, there is precedent with cffi.

If you pursue this, I would be excellent to add an example (probably separate from [the existing one](https://github.com/mdavidsaver/setuptools_dso/tree/master/example)), which would also act as integration test during CI builds.  Like what is being done currently:

https://github.com/mdavidsaver/setuptools_dso/blob/6b1b72f012bdba1c891ce7719dbd230345924209/travis-docker.sh#L22-L32

It appears that `setup_requires=` can be redirected to a local "repository".

```
[easy_install]
allow_hosts = ''
find_links = file:///path/to/local/archives/
```

cf. https://pip.pypa.io/en/latest/reference/pip_install/#controlling-setup-requires

**Comment by mdavidsaver** [comment]
I've reworked travis-docker.sh and the dsodemo example to make use of pyproject.toml.  I think this would also make it easier to incorporate a second example.

**Comment by ktbarrett** [comment]
Does this need to stay open? There is documentation on the subject now ([here](https://mdavidsaver.github.io/setuptools_dso/usage.html#pyproject-toml)).
